src/EditorFeatures/Core.Wpf/Options/EditorConfigDocumentOptionsProvider.cs

```diff
@@ -46,23 +79,69 @@ private void Workspace_DocumentClosed(object sender, DocumentEventArgs e)
                     _openDocumentContexts.Remove(e.Document.Id);
 
                     // Ensure we dispose the context, which we'll do asynchronously
-                    contextTask.ContinueWith(
-                        t => t.Result.Dispose(),
-                        CancellationToken.None,
-                        TaskContinuationOptions.OnlyOnRanToCompletion,
-                        TaskScheduler.Default);
+                    EnsureContextCleanup(contextTask);
                 }
             }
         }
 
-        private void Workspace_DocumentOpened(object sender, DocumentEventArgs e)
+        private void OnDocumentOpened(object sender, DocumentEventArgs e)
         {
             lock (_gate)
             {
-                _openDocumentContexts.Add(e.Document.Id, Task.Run(() => GetConventionContextAsync(e.Document.FilePath, CancellationToken.None)));
+                var contextTask = Task.Run(async () =>
+                {
+                    var context = await GetConventionContextAsync(e.Document.FilePath, CancellationToken.None).ConfigureAwait(false);
+                    context.CodingConventionsChangedAsync += OnCodingConventionsChangedAsync;
+                    return context;
+                });
```

Comment:
Note: wrt testing, all this async work is untracked.  that's generally not something we like to have as it means we can't test deterministically.   If we did want any testing here (which seems desirable given the bugs we're trying to fix), we'd need AsyncTokens threaded through this stuff. 

Note: i think we definitely want tests here.  This is a pretty important scenario, and it would be very easy to break it in the future.  So we shoudl design this part to be testable and write real tests that ensure that if we do change coding conventions that both teh workspace/solution gets updated, and every dependent feature now does the right thing with the new editor config settings.

Reply:
sure. but that's existing pattern. I am not fixing that in this PR. we can open tracking bug but with new editorconfig work coming from compiler, I dont think we need that since this will all change once that work is checked in.

Reply:
> I am not fixing that in this PR. 

How is this intended to be tested then?  Presumably we will have tests for this in this PR?  I'm just pointing out that testing this is currently going to be very difficult because you're doing async work in a way that can't be tracked by our testing harness.

Reply:
current test coverage is integration tests. opened issue saying we want unit test coverage as well - https://github.com/dotnet/roslyn/issues/26396

Reply:
Does the integration test simulate the files changing on disk, and us getting notified and reloading the solution?  Can you point me to the test?

Reply:
nope. we don't have a specific test for it as a lot of VS specific behavior isn't there in test. test is done manually.